### ISSUE TITLE
Install troubles : ImportError (dynamic module...

### ISSUE THREAD

**Issue by elbixos** [nest/nest-simulator#892]
Hi, and thansk for proposing NEST.
I had the following problem when trying to install. It seems that the compilation went ok but the tests say : 

Phase 7: Running PyNEST tests.
------------------------------

  Using nosetests.

  Failure: ImportError (dynamic module does not define init function (initpynestkernel)) ... ERROR

Here are the reports of the install

[reports.tar.gz](https://github.com/nest/nest-simulator/files/1737159/reports.tar.gz)

**Comment by terhorstd** [comment]
There seems to be something wrong in your python setup. The error message mentions both versions
```
======================================================================
ERROR: Failure: ImportError (dynamic module does not define init function (initpynestkernel))
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/lib/python2.7/dist-packages/nose/loader.py", line 418, in loadTestsFromName
    addr.filename, addr.module)
  File "/usr/lib/python2.7/dist-packages/nose/importer.py", line 47, in importFromPath
    return self.importFromDir(dir_path, fqname)
  File "/usr/lib/python2.7/dist-packages/nose/importer.py", line 94, in importFromDir
    mod = load_module(part_fqname, fh, filename, desc)
  File "/usr/bin/lib/python3.5/site-packages/nest/__init__.py", line 70, in <module>
    from . import pynestkernel as _kernel      # noqa
ImportError: dynamic module does not define init function (initpynestkernel)
```
Try to sort out which python version you want to use. Run *either* 2.7 *or* 3.5, but not both :)

**Comment by elbixos** [comment]
Ok, I definitely have python 2.7 and python 3.5 installed on my pc.
My last attempt also failed and here is what I've done :

1. I started from scratch with a new clone of the github repository.
2. here is my command line for cmake 
cmake -DCMAKE_INSTALL_PREFIX=/usr/bin \
	-Dwith-python=3 \
        -DPYTHON_EXECUTABLE=/usr/bin/python3 \
        -DPYTHON_LIBRARY=/usr/lib/x86_64-linux-gnu/libpython3.5m.so \
        -DPYTHON_INCLUDE_DIR=/usr/include/python3.5 \
        -DPYTHON_INCLUDE_DIR2=/usr/include/x86_64-linux-gnu/python3.5 \
/home/vpage/Nest/nest-simulator

3. Then make; sudo make install

4. make installcheck 
and same error.
Did I missed something ?


Anyway, I'm going to start a brand new virtual machine with only python 3 and see how it goes.

**Comment by elbixos** [comment]
Ok, brand new VM ubuntu 16.04.
Uninstalled python2.7

installed a bunch of packages :

sudo apt-get install build-essential python3 libreadline6 libreadline6-dev numpy python3-numpy python3-scipy python3-matplotlib cmake libgsl-dev libltdl-dev doxygen

Note that I forgot python3-nose




first Attempt :
---------------------------------------------------------------------------------------
cmake -DCMAKE_INSTALL_PREFIX=/usr/bin -Dwith-python=3 .

Failure : 
CMake Error at /usr/share/cmake-3.5/Modules/FindPackageHandleStandardArgs.cmake:148 (message):
  Could NOT find PythonInterp: Found unsuitable version "1.4", but required
  is at least "3" (found /usr/bin/python)
---------------------------------------------------------------------------------------


second Attempt :
---------------------------------------------------------------------------------------
cmake -DCMAKE_INSTALL_PREFIX=/usr/bin \
> -DPYTHON_EXECUTABLE=/usr/bin/python3 \
> -DPYTHON_LIBRARY=/usr/lib/x86_64-linux-gnu/libpython3.5m.so \
> -DPYTHON_INCLUDE_DIR=/usr/include/python3.5 \
> -DPYTHON_INCLUDE_DIR2=/usr/include/x86_64-linux-gnu/python3.5 \
> /home/vpage/nest-simulator-master/

Cmake seems ok excepet
CMake Warning:
  Manually-specified variables were not used by the project:

    PYTHON_INCLUDE_DIR2

Then make :
Everything went smoothly until
[ 99%] Compiling Cython CXX source for pynestkernel...
warning: pynestkernel.pyx:493:10: Unreachable code
warning: pynestkernel.pyx:493:10: Unreachable code
Scanning dependencies of target pynestkernel
[ 99%] Building CXX object pynest/CMakeFiles/pynestkernel.dir/pynestkernel.cxx.o
/home/vpage/nest-simulator-master/pynest/pynestkernel.cxx: In function ‘PyObject* __pyx_f_12pynestkernel_sli_array_to_object(ArrayDatum*)’:
/home/vpage/nest-simulator-master/pynest/pynestkernel.cxx:8233:35: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]
     for (__pyx_temp=0; __pyx_temp < __pyx_v_dat->size(); __pyx_temp++) {
                                   ^
/home/vpage/nest-simulator-master/pynest/pynestkernel.cxx:8259:33: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]
   for (__pyx_t_3 = 0; __pyx_t_3 < __pyx_t_2; __pyx_t_3+=1) {
                                 ^
[100%] Linking CXX shared module pynestkernel.so
[100%] Built target pynestkernel

but it seems ok...
sudo make install goes ok.

then
make installcheck
Which goes better than previously 

Nosetests unavailable Using fallback test harness.
  Fewer tests will be excuted....

but it goes through some tests
  PyNEST tests: 277
     Passed: 270
     Skipped: 7
     Failed: 0

---------------------------------------------------------------------------------------
Third Attempt 
sudo apt-get install python3-nose

and
make installcheck

same results
Nosetests unavailable

It looks like it works.
I'll try to develop a few tests on my own tomorrow.

---------------------------------------------------------------------------------------

**Comment by elbixos** [comment]
Ok, another install on a different machine, this time using
- virtualenv to select python3
- pip for installing python modules.

Everything seems ok, even the nose part that failed before.

I would recommend this solution. I'll try it on the original pc in a few days.

**Comment by richard20020815** [comment]
@elbixos 
i used fink(finkproject.org)to create an Linux environment on my Mac
using virtualenv will cause Nest to say "Could not find Pythoninterp .....find 2.7.14 but....." when cmake

so i tried your
-DPYTHON_EXECUTABLE=/usr/bin/python3 
-DPYTHON_LIBRARY=/usr/lib/x86_64-linux-gnu/libpython3.5m.so 
-DPYTHON_INCLUDE_DIR=/usr/include/python3.5 

but i can't find the .so file
so i changed it to the libpython3.6.dylib(i don't know what does .so or .dylib means...i just thought the two files' name looked similar

cmake and make install passed smoothly
on installcheck matplotlib complains that python is not installed as a framework

but whatever....
....IN PYTHON3 IMPORT NEST PASSED!!!
AFTER NEARLY A MONTH'S TRYING!!!

Depending on the result your solution works...but does that matplotlib warning mean something?

**Comment by terhorstd** [comment]
fixed by PR #984
